cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,923 @@
+    Using batch querying to avoid multiple database roundtrips.
+
+    Args:
+        component_type: The component type to check against
+        base_slug: The base slug to make unique
+        max_attempts: Maximum number of attempts to prevent infinite loops
+
+    Returns:
+        A unique slug string
+
+    Raises:
+        RuntimeError: If unable to find unique slug within max_attempts
+    """
+    if not component_type:
+        base_key = base_slug
+    else:
+        base_key = f"{component_type}:{base_slug}"
+
+    existing_publishable_entity_keys = context.get_migrated_entities(base_key)
+
+    # Check if base slug is available
+    if base_key not in existing_publishable_entity_keys:
+        return base_slug
+
+    # Try numbered variations until we find one that doesn't exist
+    for i in range(1, max_attempts + 1):
+        candidate_slug = f"{base_slug}_{i}"
+        candidate_key = f"{component_type}:{candidate_slug}" if component_type else candidate_slug
+
+        if candidate_key not in existing_publishable_entity_keys:
+            return candidate_slug
+
+    raise RuntimeError(f"Unable to find unique slug after {max_attempts} attempts for base: {base_slug}")
+
+
+def _get_title_from_olx(olx_str: str) -> str | None:
+    """
+    Extract the title from an OLX string.
+
+    Args:
+        olx_str: The OLX content string to extract title from
+
+    Returns:
+        The extracted title string
+    """
+    try:
+        olx_node = etree.fromstring(olx_str)
+        return blocks._title_from_olx_node(olx_node)
+    except (etree.XMLSyntaxError, AttributeError, TypeError) as e:
+        log.debug(f"Failed to parse OLX for title extraction: {e}")
+    except Exception as e:
+        log.warning(f"Unexpected error during title extraction: {e}")
+
+
+def _slugify_source_usage_key(
+    key: UsageKey,
+    title: str,
+    use_hash: bool = False
+) -> str:
+    """
+    Return an appropriate slug (aka block_id, aka container_id) for the target entity.
+
+    When preserve_url_slugs is False, attempts to create a human-readable slug
+    from the content title, falling back to the original block_id if parsing fails.
+    Otherwise, returns the original block_id.
+
+    Args:
+        key: The source usage key
+        olx_str: The OLX content string to extract title from
+        preserve_url_slugs: Whether to preserve original block IDs
+
+    Returns:
+        A slug string suitable for use as a block identifier
+    """
+    # Slugify the title and append some random characters for uniqueness
+    slug = slugify(title, allow_unicode=True)
+    if slug:
+        if use_hash:
+            return f"{slug}-{uuid4().hex[-6:]}"
+        return slug
+
+    # Fallback to original block_id
+    return key.block_id
+
+
+# Not used for now since it causes issues with
+# adding to Collections
+# Required for additional investigation
+def _create_migration_artifacts_in_bulk(
```

Comment:
If we're going to leave this in the PR, then please create follow-up issue and link it here to help prevent us from forgetting about it and leaving dead code here.

Reply:
Can't reproduce the issue. Will test additional use cases.
@kdmccormick What do you think will be better - to leave two different function implementing a third coordinator func to invoke a desired logic based on input param (e.g. create_in_bulk=True) and simple leave one option.
I'm not sure on the bulk option since it's really complicated and obviously harder to maintain than incremental one. And the performance boost for regular courses is not so huge.